File: packages/evals/tasks/bench/extract/extract_csa.ts
import { z } from "zod";
import { defineBenchTask } from "../../../framework/defineTask.js";
import { normalizeString } from "../../../framework/textScoring.js";

export default defineBenchTask(
  { name: "extract_csa" },
  async ({ debugUrl, sessionUrl, stagehand, page, logger }) => {
    try {
      await page.goto("https://browserbase.github.io/stagehand-eval-sites/sites/csa/");

      const { data: result } = await stagehand.extract(
        "Extract all the publications on the page including the publication date, session type, publication type, and annotation",
        z.object({
          publications: z.array(
            z.object({
              publication_date: z.string(),
              session_type: z.string(),
              publication_type: z.string(),
              annotation: z.string(),
            }),
          ),
        }),
      );

      const publications = result.publications;
      const expectedPublications = await page.evaluate(() =>
        Array.from(document.querySelectorAll("table tbody tr")).flatMap((row) => {
          const cells = Array.from(row.querySelectorAll(":scope > td"));
          if (cells.length !== 5) return [];
          return [
            {
              publication_date: cells[0]?.textContent?.trim() ?? "",
              session_type: cells[2]?.textContent?.trim() ?? "",
              publication_type: cells[3]?.textContent?.trim() ?? "",
              annotation: cells[4]?.textContent?.trim() ?? "",
            },
          ];
        }),
      );
      const key = (publication: {
        publication_date: string;
        session_type: string;
        publication_type: string;
        annotation: string;
      }) =>
        [
          normalizeString(publication.publication_date),
          normalizeString(publication.session_type),
          normalizeString(publication.publication_type),
          normalizeString(publication.annotation),
        ].join("|");
      const actualKeys = publications.map(key).sort();
      const expectedKeys = expectedPublications.map(key).sort();
      const allPublicationsMatch =
        actualKeys.length === expectedKeys.length &&
        JSON.stringify(actualKeys) === JSON.stringify(expectedKeys);

      if (!allPublicationsMatch) {
        logger.error({
          message: "Extracted publications do not match the page",
          level: 0,
          auxiliary: {
            expected: { value: JSON.stringify(expectedPublications), type: "object" },
            actual: { value: JSON.stringify(publications), type: "object" },
          },
        });
        return {
          _success: false,
          error: "Extracted publications do not match the page",
          logs: logger.getLogs(),
          debugUrl,
          sessionUrl,
        };
      }

      return {
        _success: true,
        logs: logger.getLogs(),
        debugUrl,
        sessionUrl,
      };
    } catch (error) {
      return {
        _success: false,
        error: String(error),
        logs: logger.getLogs(),
        debugUrl,
        sessionUrl,
      };
    }
  },
);

File: packages/sdk-ruby/evals/tasks/extract/extract_csa.rb
# frozen_string_literal: true

# Port of packages/evals/tasks/bench/extract/extract_csa.ts

Evals.define_task("extract_csa") do |t|
  t.page.goto("https://browserbase.github.io/stagehand-eval-sites/sites/csa/")

  result = t.stagehand.extract(
    "Extract all the publications on the page including the publication date, " \
    "session type, publication type, and annotation",
    schema: {
      "type" => "object",
      "properties" => {
        "publications" => {
          "type" => "array",
          "items" => {
            "type" => "object",
            "properties" => {
              "publication_date" => { "type" => "string" },
              "session_type" => { "type" => "string" },
              "publication_type" => { "type" => "string" },
              "annotation" => { "type" => "string" },
            },
            "required" => %w[publication_date session_type publication_type annotation],
            "additionalProperties" => false,
          },
        },
      },
      "required" => ["publications"],
      "additionalProperties" => false,
    },
    page: t.page,
  )

  publications = result.data["publications"]
  expected_length = 14

  expected_first_item = {
    "publication_date" => "11-30-2024",
    "session_type" => "Regular Session",
    "publication_type" => "Assembly Weekly History",
    "annotation" =>
      "2024 -- This publication includes the complete histories of second-year bills. " \
      "The complete electronic history of all bills is always available at leginfo.legislature.ca.gov",
  }

  expected_last_item = {
    "publication_date" => "11-30-2016",
    "session_type" => "1st Extraordinary Session",
    "publication_type" => "Assembly Weekly History",
    "annotation" => "",
  }

  if publications.length < expected_length
    t.logger.error("Incorrect number of publications extracted",
                   { expected: ">= #{expected_length}", actual: publications.length })
    next { _success: false, error: "Incorrect number of publications extracted" }
  end

  has_expected_first_item = publications.any? do |publication|
    publication["publication_date"] == expected_first_item["publication_date"] &&
      publication["session_type"] == expected_first_item["session_type"] &&
      publication["publication_type"] == expected_first_item["publication_type"] &&
      publication["annotation"] == expected_first_item["annotation"]
  end

  unless has_expected_first_item
    t.logger.error("Expected 'first' item not found in publications",
                   { expected: expected_first_item, actual: publications })
    next { _success: false, error: "Expected 'first' item not found in publications" }
  end

  has_expected_last_item = publications.any? do |publication|
    publication["publication_date"] == expected_last_item["publication_date"] &&
      publication["session_type"] == expected_last_item["session_type"] &&
      publication["publication_type"] == expected_last_item["publication_type"] &&
      publication["annotation"] == expected_last_item["annotation"]
  end

  unless has_expected_last_item
    t.logger.error("Expected 'last' item not found in publications",
                   { expected: expected_last_item, actual: publications })
    next { _success: false, error: "Expected 'last' item not found in publications" }
  end

  { _success: true }
end
